libvips/create/text.c

```diff
@@ -389,15 +409,15 @@ vips_text_build( VipsObject *object )
 		return( -1 );
 	}
 
-	text->fontmap = pango_cairo_font_map_get_default();
```

Comment:
It used to be the case that fontmap would not unref cleanly on some backends (win32? coretext? I forget), so it was important to keep them as long as possible and reuse them as much as possible. Perhaps this has been improved now? It was a while ago. 

If it hasn't been improved, then I think we should probably have a single fontmap we share between all instances of `text`.

Another alternative would be to scale text only in cairo rather than by changing dpi on the fontmap. This should be faster and more threadsafe, though I wonder if quality might be lower --- I would guess changing dpi is more likely to get the right hints sent down to the font rendering system.

Reply:
Ah, I was unaware of that unref bug. Reusing the fontmap sounds OK in that case. I also wasn't sure how expensive the initial call to `pango_cairo_font_map_new()` would be, so perhaps a lock is faster than using a `GPrivate`.

Scaling via Cairo would likely reduce quality, which indeed should be handled by its font rendering backend (win32, Core Text, Fontconfig, etc.).

I think the issue with wxWidgets is that it does not use [`gtk_widget_set_font_map()`](https://docs.gtk.org/gtk4/method.Widget.set_font_map.html), so it would always end-up using the fontmap from `pango_cairo_font_map_get_default()`, see e.g.:
https://github.com/wxWidgets/wxWidgets/blob/662e23b432773515568224d00bf5baf269f5284a/src/gtk/window.cpp#L373
https://github.com/GNOME/gtk/blob/071c6c60c0784674bc52daf48fb0a94c19d28754/gtk/gtkwidget.c#L6400
https://github.com/GNOME/gtk/blob/071c6c60c0784674bc52daf48fb0a94c19d28754/gtk/gtkwidget.c#L6608

Reply:
IIUC (but correct me if I'm wrong), the `*_build` functions are rarely called concurrently on different threads (it's usually only called on the main thread). So, perhaps it does not matter so much whether you use a `GPrivate` or a lock here.

Let me know if I need to check the timing of calling `pango_cairo_font_map_new()`,  I suppose if that's an expensive one we'd better just reuse it with locking (i.e. this approach).

Reply:
You can have several main threads, so `_build()` can run concurrently. I don't know how common this is though. 

The unref problems with `pango_cairo_font_map_new()` certainly used to be serious (large scale memory leaks with some backends), so I think it's probably best to just have one. Though that was a few years ago now, maybe these issues have been fixed? We'd need to do some testing to be sure.

The testing would be a PITA and the benefits of more threading in `text` seem small, so my 2p would be to just have one font map.

Reply:
I agree, let's just have one font map.